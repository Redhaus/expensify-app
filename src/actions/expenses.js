import uuid from 'uuid';

// ACTIONS

// ADD_EXPENSE
// create an action generator with implicit return,  creating a const with deconstructred items and their 
// defaults 
// this action has type of ADDEXPENSE 
// With data sent as expense, found under action.expense in reducer
// this action accepts an object as params
export const addExpense = ( 
    {
        description = '', 
        note = '', 
        amount = 0, 
        createdAt = 0 
    } = {} 
) => ({
    type: 'ADD_EXPENSE',
    expense:{
        id: uuid(),
        description: description,
        note: note, 
        amount: amount,
        createdAt: createdAt

    }
})





// REMOVE_EXPENSE
// Set up the action generator
export const removeExpense = ( { id } = {} ) => ({
    type: 'REMOVE_EXPENSE',
    id: id
    
})



// EDIT_EXPENSE

export const editExpense  = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id: id,
    updates: updates


})