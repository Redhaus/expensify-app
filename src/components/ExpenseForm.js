import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



class ExpenseForm extends React.Component {

   constructor(props) {
       super(props);

       const exp = props.expense;

       this.state = {
        description: exp ? exp.description : '',
        note: exp ? exp.note : '',
        amount: exp ? (exp.amount / 100).toString() : '',
        createdAt: exp ? moment(exp.createdAt) : moment(),
        calFocused: false,
        error: ''

    }
       
   }
   
    
    

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( (state) => ({description: description}) )

    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState( (state) => ({note: note}))
    }

    onAmountChange = (e) => {

        // match function makes sure var matches regular expression
        // regular exp says only digets as many as you want, 
        // then a decimal and 2 more digits are optional
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
         this.setState((state) => ({amount: amount}))
        
        }


    }

    handleOnSubmit = (e) => {
        e.preventDefault();

        const description = this.state.description;
        const amount = this.state.amount;
        const note = this.state.note;
        const createdAt = this.state.createdAt;

        if(!description || !amount){
            //Set Error State equal to please provide date and amount
            this.setState( () => ({ error: 'Please Provide Date and Amount'}))
        }else{
            console.log('submitted');
            this.setState( () => ({ error: ''}))

            // use onSubmit prop function and pass in state data
            this.props.onSubmit({

                description: description,
                amount: parseFloat(amount, 10) * 100,
                createdAt: createdAt.valueOf(),
                note: note

            })
            
        }

        

    }

    onDateChange = (selecteddate) => {
        if (selecteddate){
        this.setState( () => ({createdAt: selecteddate}))
        }
    }

    onFocusChange = ( {focused}) => {
        this.setState( () => ({ calFocused: focused }))
    }

    render(){
        return (
            <div>
                Expense Form
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.handleOnSubmit}>
                <input 
                    type="text" 
                    name="description"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input 
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={ () => false }
                
                />

                <textarea 
                    placeholder="Add note for expense..."
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    
                >
                </textarea>
                <button type="submit">{this.props.expense ? 'Update Expense' : 'Add Expense' }</button>
                </form>

            </div>
        )
    }
}

export default ExpenseForm;