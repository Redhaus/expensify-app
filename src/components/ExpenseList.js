

import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import ExpenseListFilters from './ExpenseListFilters';

// map through number of expenses and display expenseListItem passing unique ID and spread expense object
export const ExpenseList = (props) => (
    <div>
       
        <ExpenseListFilters />
        
        
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                props.expenses.map( (expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />     
                })
            )
        }
        

   
        
    </div>
)



// makes redux store / state to various props
// return what you want from the state and then access it via props
// define items we want to get off redux state
// i.e. props.expenses and props.filters
// using select expenses you are adding filtering into expenses props object
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

// const for HOC similar, connect returns a function, 
// like the function we created for HOC's pass connect return the component
// supply defined items you want as props, and provide wrapped component
export default connect(mapStateToProps)(ExpenseList);
