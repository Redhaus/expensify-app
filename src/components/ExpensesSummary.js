// Create a summary component

// rendered by ExpenseDashboard Page
// Test with two snapshots
// Connected to store for:
//     expenseCount - how many visible expenses - selector
//     expensesTotal - whats the total of visible expenses - selector


// tests 
// Example viewing 2 expenses totalling 94.34
// example viewing 1 expense totalling 98.30

// deploy

// Commit and Deploy

// Get the feature live

import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expensesTotal';
import numeral from 'numeral';





export const ExpenseSummary = ({expenseCount, expenseTotal}) => {

    console.log(expenseCount, expenseTotal);

    if(!expenseCount){
        return ''

    }else{

        const expTitle = expenseCount > 1 ? 'expenses' : 'expense';

        return(
            <div>
                <h3>Viewing {expenseCount} {expTitle} totaling {expenseTotal} dollars.</h3>
            </div>
        )
    }
}


const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    const total = getExpensesTotal(visibleExpenses);

    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: numeral(total / 100).format('$0,0.00')
    }

}


export default connect(mapStateToProps)(ExpenseSummary);






// export class ExpenseSummary extends React.Component {

//     getTotals = (expense) => {
//         const total = getExpensesTotal(expense)
//         return numeral(total / 100).format('$0,0.00');
//     }



//     render(){
//     console.log(this.props.expenses);
    
//         return(
            
//             <div>{this.props.expenses.length > 0 
//                     ? `Viewing ${this.props.expenses.length} totaling ${this.getTotals(this.props.expenses)}` 
//                     : ''}
//             </div>
//         )

//     }
    

// }

// const mapStateToProps = (state) => {
//     // const visibleExpenses = 
//     return {
//         expenses: selectExpenses(state.expenses, state.filters),
//     }
// }

// // const mapDispatchToProps = (dispatch) => {
// //     return {
// //     getExpensesTotal: (data) => dispatch( getExpensesTotal(data) )
// //     }
// // }

// export default connect(mapStateToProps)(ExpenseSummary);

