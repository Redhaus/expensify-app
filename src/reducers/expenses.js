
// REDUCERS


// default setup for initial value of state
const expensesReducerDefaultState = []

// Expenses Reducer
// reducer accepts state you define and action 
// Add_expense will return new array of existing state plus expense from action function

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type){
        case 'ADD_EXPENSE':
            return [...state, action.expense]
//filter through every item in array, if items id, does not = action expense id 
// keep it in the array, if it matches remove item in array
        case 'REMOVE_EXPENSE':
            // console.log(action.expense.id);
            return state.filter( (item) => item.id !== action.id )
        case 'EDIT_EXPENSE':
            ///map through every item in state
            // if item.id matches the action.id 
            // spread out item with match id overwrite with updates 
            return state.map( (item) => {
                if(item.id === action.id){
                    return {
                        ...item, 
                        ...action.updates
                        
                    }

                }else{
                    return item
                }
            } )
        default:
            return state;
    }
}

export default expensesReducer;