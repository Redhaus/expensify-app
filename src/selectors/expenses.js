import moment from 'moment';
// timestamps (milliseconds)
// Jan 1st 1970 (unix epoch)
// positive numbers are forward, neg numbrs and before that

// GET VISIBLE EXPENSES
// set up visible function that accepts expenses and filter objects
// deconstruct filter object 
// filter tthrough each expense 
// create a const for major filters start, end and text

// If all are true that object item remains in filter
// If any test is not true it is removed

// startDateMatch is test to see if provided if not provided it is undefined 
// and come back true

// If it is defined check to see if it comes after created by date
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter( (expense) => {
        const createdAtMoment = moment(expense.createdAt)

        // set up a expense description to lowercase and assign as var
        const description = expense.description.toLowerCase();
        // set search term to lower case for comparison
        const search = text.toLowerCase();
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = description.includes(search); // if this returns true with other two result is gathered.

        return startDateMatch && endDateMatch && textMatch

    // sort will always be a defaulted filter option to date or amount
    // so chain the sort function onto the filter function and returns 
    // each item is compared and evaluated to be up in order
    }).sort( (a,b) => {
        if(sortBy==='date'){
            return a.createdAt < b.createdAt ? 1 : -1
        }else{
            console.log('sort called');
            
            return a.amount < b.amount ? 1 : -1
        }
    });
}


export default getVisibleExpenses