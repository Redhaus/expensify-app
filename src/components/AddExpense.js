import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import { addExpense } from '../actions/expenses';


export class AddExpensePage extends React.Component {

    onSubmit = (expense)=> {
        // props.dispatch( addExpense(expense));
        this.props.addExpense(expense);
        this.props.history.push('/');
    }

    render() {
        return (
        <div>
        <h3>Add Expense</h3>
        <ExpenseForm 
            onSubmit={ this.onSubmit } 
        />
    </div>
    )
    }
}

//create onSubmit prop unction to be passed down that passes expense object state data passed up to it.
//call the props dispatch after connecting component and call addExpense func and pass it expense 


const mapDispatchToProps = (dispatch)  => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);






// WORKING VERSION 
// import React from 'react';
// import ExpenseForm from './ExpenseForm';



// import { connect } from 'react-redux';
// import { addExpense } from '../actions/expenses';


// const AddExpensePage = (props) => {


//     const handleAddExpense = (e) => {
//         e.preventDefault();
//         const description = e.target.description.value;
//         const amount = e.target.amount.value;
//         const date = e.target.date.value;
//         const note = e.target.note.value
//         props.dispatch( addExpense({
//             description: description, 
//             amount: amount, 
//             date: date,
//             note: note
//         }));

//         e.target.description.value = '';
//         e.target.amount.value = '';
//         e.target.date.value = '';
//         e.target.note.value = '';
//     }

//     return (
//     <div>
//         <h3>Add Expense</h3>
//         <form onSubmit={handleAddExpense}>
        
//         <div>
//         <label>Description</label>
//         <input type="text" name="description"/>
//         </div>

//         <div>
//         <label>Amount</label>
//         <input type="text" name="amount"/>
//         </div>
        
//         <div>
//         <label>Date</label>
//         <input type="text" name="date"/>
//         </div>
        
//         <div>
//         <label>Note</label>
//         <input type="textarea" name="note"/>
//         </div>

//         <button type="submit">Add Expense</button>
//         </form>
//     </div>
//     )
// }



// export default connect()(AddExpensePage);