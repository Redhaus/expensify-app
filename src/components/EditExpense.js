



///
import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';
import { removeExpense } from '../actions/expenses'


// refactor to be class based component
// setup map dispatch to props for editExpense and removeExpense

//should render editExpense Page with snapshot
//should handle editExpense using spies
// should handle remove expense using spies 

export class EditPage extends React.Component{

    handleRemoveExpense = () => {
        // this.props.dispatch( removeExpense({ id: this.props.expense.id}) )
        this.props.removeExpense({ id: this.props.expense.id}) 
        this.props.history.push('/');
        
    }

    onSubmit = (expense) => {
        this.props.editExpense( this.props.expense.id, expense );
        this.props.history.push('/');

    }

    render(){
        return(
            <div>
            <ExpenseForm 
                expense={this.props.expense}
                onSubmit={ this.onSubmit }
            />
             <button onClick={this.handleRemoveExpense}>Remove</button>
             </div>  
        )
    }

}


// you can pass state and props to mapStateToProps
// since you have the id of the individual expense page we are on 
// from the array 

const mapStateToProps = (state, props) => {
    const id = props.match.params.id;
    return {
        expense: state.expenses.find( (expense) => {
            return expense.id === id;
        })
    }

}

const mapDispatchToProps = (dispatch, props) => {
    return {
        removeExpense: (data) => dispatch( removeExpense({ data }) ),
        editExpense: (id, expense) =>  dispatch( editExpense( id, expense ))

    }

}

export default connect(mapStateToProps, mapDispatchToProps)(EditPage);