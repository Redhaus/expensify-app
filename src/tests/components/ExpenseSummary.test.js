import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import {ExpenseSummary} from '../../components/ExpensesSummary';
import React from 'react';




test('TEST NO EXPENSE', () => {
    const wrapper = shallow(<ExpenseSummary/>);
    expect(wrapper).toMatchSnapshot(); 
})

test('TEST MULTIPLE EXPENSE', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={2} expenseTotal={345}/>);
    expect(wrapper).toMatchSnapshot(); 
})

