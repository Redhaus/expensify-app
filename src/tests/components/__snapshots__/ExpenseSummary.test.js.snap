// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST MULTIPLE EXPENSE 1`] = `
<div>
  <h3>
    Viewing 
    2
     
    expenses
     totaling 
    345
     dollars.
  </h3>
</div>
`;

exports[`TEST NO EXPENSE 1`] = `""`;
